common/src/main/java/org/opensearch/ml/common/FunctionName.java

```diff
@@ -35,7 +36,7 @@ public static FunctionName from(String value) {
      * @return true for deep learning model.
      */
     public static boolean isDLModel(FunctionName functionName) {
-        if (functionName == TEXT_EMBEDDING || functionName == SPARSE_ENCODING || functionName == SPARSE_TOKENIZE) {
+        if (functionName == TEXT_EMBEDDING || functionName == SPARSE_ENCODING || functionName == SPARSE_TOKENIZE || functionName == TEXT_SIMILARITY) {
```

Comment:
why we choose a name called as TEXT_SIMILARITY?

Reply:
isn't this a re-ranking?

Reply:
+1 on this.

Reply:
I chose `TEXT_SIMILARITY` because I think that best captures what this algorithm is about. Sure, we use it for Reranking, but you can use it for other stuff, too. For instance, one thing we've been using cross-encoders for is computing the semantic similarity between answers generated by our RAG systems and human-annotated answers that we know are correct, which is not a reranking use-case. Of course, we don't do that on the cluster, so maybe it's moot, but I wanted to pick a name that captures all use-cases of cross-encoders (or any models that spit out numbers representing how similar two pieces of text are). wdyt?

Reply:
I understand your point. But for semantic search, we also perform the "text similarity" with comparing large numbers of texts in our vector store. So IMO, text similarity can be achieved through bi-encoder & cross encoder both. In fact, with the efficiency concern bi-encoder is widely used for text similarity. 

But for Re-ranking cross encoder is widely used. So I would vote to rename this to `RE-RANK` algorithm and customer can choose which algorithm to choose to obtain their text similarity goal.

Reply:
I  guess my concern with calling it `RERANK` is that there are also other reranking interfaces and models - not all are just reranking text. e.g. more traditional things like xgboost. idk. probably it's fine. maybe `TEXT_COMPARISON`?

Reply:
> But for Re-ranking cross encoder is widely used. So I would vote to rename this to RE-RANK algorithm and customer can choose which algorithm to choose to obtain their text similarity goal.

When I proposed cross-encoder reranking to the neural search folks their main feedback was that reranking should be generalizable to more kinds of models than cross-encoders (see [this rfc](https://github.com/opensearch-project/neural-search/issues/485)), so I'm wary of synonymizing reranking with cross-encoders

Reply:
I'm not seeing any issue honestly. I would like to see this way: Currently we are going to support cross-encoder for re-ranking. We can extend `re-ranking` with other mechanism later when we have more clear requirement.

@navneet1v what do you think? Do you see any issue?

Reply:
@navneet1v unless I'm misunderstanding, this enum is meant to be 1:1 with the model interface. So for example you can swap out MiniLM for openai-ada (local vs remote) within the bounds of `TEXT_EMBEDDING`, because both models take a list of strings and give back a list of vectors. 

This is an interface that takes a string and a list of strings and gives back a list of numbers (representing the similarity of the first string with list, though I guess the interface doesn't demand that). You could just as easily replace the cross encoder with something like a BLEU score (although you'd need to implement that algorithm) and not break the contract.

@dhrubo-os is that a correct understanding of FunctionName?

Reply:
In my opinion, this is 1:N relationship with different model types. For example, for text embedding we are supporting multiple sentence embedding model types, like sentence transformers model, clip models. 

Let's take an example of `REMOTE`. This interface is supporting multiple model types including multiple LLMs as a remote model. 


@ylwu-amzn please correct me if I'm wrong.

Reply:
The function name is to group models with same interface and task types. Maybe some models for different tasks have same interface, but it maybe confused if we group them to one enum. I would suggest name for general task purpose. @HenryL27 , this is a long discussion, can you summarize what options we have now and explain the reason why people prefer the name?

Reply:
sure. summary:
We're debating the name of the function that cross-encoders should use. Options:

1. `TEXT_SIMILARITY` (original choice): This is a model that measures the similarity of text. ergo, text similarity.
2. `RE_RANK`: This is a model used for reranking. ergo, re-rank
3. `TEXT_COMPARISON`: This is a model that compares pieces of text (similarity can be measured by vector distance and is therefore too broad. Although I guess an implementation of this algorithm that uses a bi-encoder and vector similarity is just as valid as the cross-encoder implementation), but the output can mean whatever you want it to mean, it's just a number.

The drawback of `RE_RANK` is that it seems to preclude other uses of cross-encoders, and also other kinds of reranking models.

The drawback of `TEXT_*` is that is obfuscates the main use-case (re-ranking)

@dhrubo-os did I miss anything?

Reply:
@ylwu-amzn I think we just need you to make a decision here. fwiw I got chatGPT to vote too [cross encoder naming](https://platform.openai.com/playground/p/qpkL9gjOr8zKF7tDW8rzziUu?model=gpt-3.5-turbo&mode=chat)

Reply:
@dhrubo-os `re-ranking` is _not_ what the cross encoder algorithm does.  The output of the algorithm is, as you say, widely used for re-ranking purposes.

However, I agree that `text similarity` is not a uniquely identifying feature of the cross encoder.  But given the above options, I would choose 1 (1 and 3 are pretty much the same to me).  We can maybe throw in `lossless` to distinguish cross encoders from bi encoders, but I am not sure if that's 100% accurate.  I'm thinking of terms like "lossless", "HD", "hi-res", "slow", "rich".  Would adding one of those help?  Again, if we are choosing one, my vote is for 1.

Reply:
Text similarity does seem quite generic. In our semantic search implementation, we are indeed performing text similarity analysis, right? This enum refers to either a function or an algorithm name. However, my interest lies more in segregating the functionality. For instance, currently, we're achieving this through the use of a cross-encoder, but it's pertinent to note that in the future, we might explore and integrate additional methods to enhance this capability. The crucial aspect to keep in mind is that this functionality is exclusively dedicated to the re-ranking process, irrespective of the underlying method we employ to achieve it.

Think about `Question-answering` use case, In QA, text similarity algorithms might be used to match a user's question with similar questions or to find passages in a text that are similar to the question. Is that case, will we fill QA also under this giant `Text similarity` umbrella?

Reply:
@HenryL27 @dhrubo-os what's the final decision?

Reply:
`text similarity`